internal/controller/gardener_cluster_controller.go

```diff
@@ -0,0 +1,179 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	infrastructuremanagerv1 "github.com/kyma-project/infrastructure-manager/api/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	lastKubeconfigSyncAnnotation = "operator.kyma-project.io/last-sync"
+	clusterCRNameLabel           = "operator.kyma-project.io/cluster-name"
+	defaultRequeuInSeconds       = 60 * 1000
+)
+
+// GardenerClusterController reconciles a GardenerCluster object
+type GardenerClusterController struct {
+	client.Client
+	Scheme             *runtime.Scheme
+	KubeconfigProvider KubeconfigProvider
+	log                logr.Logger
+}
+
+func NewGardenerClusterController(mgr ctrl.Manager, kubeconfigProvider KubeconfigProvider, logger logr.Logger) *GardenerClusterController {
+	return &GardenerClusterController{
+		Client:             mgr.GetClient(),
+		Scheme:             mgr.GetScheme(),
+		KubeconfigProvider: kubeconfigProvider,
+		log:                logger,
+	}
+}
+
+//go:generate mockery --name=KubeconfigProvider
+type KubeconfigProvider interface {
+	Fetch(shootName string) (string, error)
+}
+
+//+kubebuilder:rbac:groups=infrastructuremanager.kyma-project.io,resources=clusters,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructuremanager.kyma-project.io,resources=clusters/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructuremanager.kyma-project.io,resources=clusters/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// the GardenerCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.0/pkg/reconcile
+func (r *GardenerClusterController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) { //nolint:revive
+	r.log.Info("Starting reconciliation loop")
+
+	var cluster infrastructuremanagerv1.GardenerCluster
+
+	err := r.Client.Get(ctx, req.NamespacedName, &cluster)
+
+	if err != nil {
+		r.log.Error(err, "could not get the CR for "+req.NamespacedName.Name)
+		return r.ResultWithoutRequeue(), err
```

Comment:
I think if we failed to read Gardener Cluster resource we should requeue the event.